WIP - Api external upgrades

respectus opened the pull request:
- Added Account Management
- Tested Account Management
- Added Initial Hippo / Chat Endpoints

Commit:
feat: add account management as a module

Commit:
added hippo endpoints, upgraded login to Personal Access Token PAT, more PAT updates to come

Commit:
added tests and upgraded hippo asks

Melabela commented on cerevox/account.py:
```diff
@@ -0,0 +1,480 @@
+            MessageResponse with revocation confirmation
+        """
+        response_data = self._request("POST", "/token/revoke")
+        return MessageResponse(**response_data)
```
Should this fn, also remove `Authorization` header from `self.session` before returning?

respectus replied:
Ah yes, I've updated this to remove the auth header from session.

Melabela commented on cerevox/async_account.py:
```diff
@@ -0,0 +1,489 @@
+        """Async context manager entry"""
+        await self.start_session()
+        # Automatically authenticate using email and password
+        await self.login(self.email, self.api_key)
```
Having this line in `__aenter__`, wouldn't it cause many login calls to API ?
- think `__aenter__` is called with every `async` block...

Should there be a check, something like "if not `token`:  then do `self.login(...)` ?

OR alternatively, move `login()` call into `start_session` (inside its `if` block), since that has a checkable class property.

respectus replied:
good catch, I've moved it into the `start_session`

Melabela commented:
Seems like there's a decent amount of repeated code/declarations in some of the new classes.  Namely:

**(sync classes - e.g. `account, hippo`)**
- `__init__()`
- `__enter__(), __exit__()`
- `_request()`
- `close()`
- Token fns:  `login(), refresh_token(), revoke_token()`

**(async classes - e.g. `async_account, async_hippo`)**
- `__init__()`
- `__aenter__(), __aexit__()`
- `start_session(), close_session()`
- `_request()`
- Token fns:  `login(), refresh_token(), revoke_token()`

The above content would really benefit from being made a Base Class, to avoid repeating/redeclaring the same content.

Melabela commented on cerevox/async_hippo.py:
```diff
@@ -0,0 +1,721 @@
+        Returns:
+            FileUploadResponse containing upload confirmation and file info
+        """
+        import os
```
(move / remove this - imports should be at top of file, esp. for standard library)

respectus replied:
removed this

Melabela commented on cerevox/hippo.py:
```diff
@@ -0,0 +1,698 @@
+        Returns:
+            FileUploadResponse containing upload confirmation and file info
+        """
+        import os
```
(move / remove this - imports should be at top of file)

respectus replied:
removed this

Melabela commented:
(Not in this PR, but now that I'm looking for it...

`document_loader.py` has several "late in file" imports:
```
document_loader.py:1424:        import string   ## [Melissa] this seems unused(?)
document_loader.py:1801:        import json  ## [Melissa] move to top-of-file
document_loader.py:2191:        import json
```
)

Melabela commented on cerevox/models.py:
```diff
@@ -294,3 +294,318 @@ class SiteListResponse(BaseModel):
+    plan: str = Field(..., description="Plan name")
+    base: int = Field(..., description="Base limit")
+    bytes: int = Field(..., description="Storage limit in bytes")
+    messages: Optional[int] = Field(None, description="Message limit")
```
_(Later)_ 
FYI, will need to update `AccountPlan` fields, after `api-external` is updated.
Believe we're not using `messages` now...

respectus replied:
Sounds good, I will clean up the models once we've finalized api-external

Melabela commented:
Scanned `(async_)account.py` and `(async_)hippo.py` - looks good.  They cover the APIs listed  in `api-external`, and the HTTP calls have the correct method & url-paths.

Melabela commented:
_(Lower priority item)_

In `test_async_account.py` and `test_async_hippo.py` there are many token-login calls (`/v1/token/login`), early in each test function.  Could these be consolidated somewhat?

Compared with (sync) `test_hippo.py`, token-login call only appears once at top of most classes, in `def setup_method(self):`

respectus commented:
> Seems like there's a decent amount of repeated code/declarations in some of the new classes. Namely:
> 
> **(sync classes - e.g. `account, hippo`)**
> 
> * `__init__()`
> * `__enter__(), __exit__()`
> * `_request()`
> * `close()`
> * Token fns:  `login(), refresh_token(), revoke_token()`
> 
> **(async classes - e.g. `async_account, async_hippo`)**
> 
> * `__init__()`
> * `__aenter__(), __aexit__()`
> * `start_session(), close_session()`
> * `_request()`
> * Token fns:  `login(), refresh_token(), revoke_token()`
> 
> The above content would really benefit from being made a Base Class, to avoid repeating/redeclaring the same content.

Good point, I will move the repeated code to base classes

respectus commented:
> (Not in this PR, but now that I'm looking for it...
> 
> `document_loader.py` has several "late in file" imports:
> 
> ```
> document_loader.py:1424:        import string   ## [Melissa] this seems unused(?)
> document_loader.py:1801:        import json  ## [Melissa] move to top-of-file
> document_loader.py:2191:        import json
> ```
> 
> )

removed the unused/ move the used to top

respectus commented:
> _(Lower priority item)_
> 
> In `test_async_account.py` and `test_async_hippo.py` there are many token-login calls (`/v1/token/login`), early in each test function. Could these be consolidated somewhat?
> 
> Compared with (sync) `test_hippo.py`, token-login call only appears once at top of most classes, in `def setup_method(self):`

I'll take a look at consolidating the tests, this part was a bit more generated so it definitely could be DRY-er. I'll look at consolidating the login functions, It'll be part of the base class refactor

Commit:
small updates: move imports to top, removed unused imports, revoke_token removes auth headers, and move login after start_session

Commit:
refactor structure to include core/ and base clients under the core/ structure

Commit:
remove openai_key param from chats

Melabela commented:
Thank you for taking the time to do the refactors.
For (sync/async) Base Classes, AND consolidating login calls in `test_async_(account/hippo).py`.

Much better on DRY now, and the code is more readable :)

Commit:
fix imports for internal files and for exposure externally

Commit:
submit_ask POST expects sources not file_sources

Commit:
updates for ask api, plus add test_live_api.py for end to end testing

Commit:
remove email, only use, key, will add login soon

Melabela commented:
Another odd import code found in `cerervox/utils/document_loader.py`:
- [Line 172](https://github.com/CerevoxAI/cerevox-python/blob/f12e66454cea792f4de33380dd6e140439f11091/cerevox/utils/document_loader.py#L172) has `pandas = importlib.import_module("pandas")`
- BUT Earlier in the file at [Line 22](https://github.com/CerevoxAI/cerevox-python/blob/f12e66454cea792f4de33380dd6e140439f11091/cerevox/utils/document_loader.py#L22), there already is `import pandas` check with `try/except` around it...
- so 1st instance (Line 172) seems redundant, and may overwrite `pandas` from file-level import (2nd Instance)
- **TL/DR:** suggest remove Line 172, and also Line 5 (`import importlib`) only used by Line 172

Melabela commented:
Also for `cerervox/utils/document_loader.py`, seems this could be split into 4 files...:
- one for small `@dataclass`'es, near top of file
- one for `class Document`, which has ~1300 lines
- one for `class DocumentBatch`, which has ~800 lines
- collection of file-level functions after above classes (last ~400 lines of file):  `chunk_markdown(), chunk_text(), _split_*(), _strip_text(), _merge_small_chunks()`

In its current "original" form, it makes sense, since the classes are strongly interrelated.
But I think with the updated `__init__.py` format of this PR, splitting the files will also work.

Melabela commented:
For `cerervox/utils/document_loader.py`, a few more suggestions:
- for fn `get_documents_by_element_type()` at [Line 1937](https://github.com/CerevoxAI/cerevox-python/blob/f12e66454cea792f4de33380dd6e140439f11091/cerevox/utils/document_loader.py#L1937):
  - rename to `filter_by_element_type()`
  - and move up in file, next to two other `filter_by_{type/page_count}()` fns with similar operation, i.e. they all return `DocumentBatch`
- fns `get_documents_without_errors()` and `get_documents_with_errors()`, similar to above... could:
  - rename prefix `get -> filter`
  - change return type `List[Document] -> DocumentBatch` for interface consistency

Melabela commented:
In `examples/cloud_integrations.py`, probably many users will only have a few Cloud Providers, i.e. subset of those we support.  So it doesn't make sense to run examples for those they don't use.

(EDIT:  just noticed `Google Drive` is not in the example..., but is listed in the Root README.md, as one of the integrations we have)

Suggestion:
- have flags near top-of-file for each cloud provider, which controls which demos runs (or not)

e.g. Add Flags near top of file
```diff
from cerevox import AsyncLexa, Lexa, LexaError, ProcessingMode

# Control which demos to run, set False if don't have that Cloud Service
+ HAVE_AMAZON_S3=True
+ HAVE_BOX=True
+ HAVE_DROPBOX=True
+ HAVE_MICROSOFT_SHAREPOINT=True
+ HAVE_SALESFORCE=True
+ HAVE_SENDME=True
```

And only run demos for Cloud Services that user has configured
```diff
    # Run examples for each cloud provider
    try:
        # Core cloud storage providers
-        amazon_s3_example(client)
-        box_example(client)
-        dropbox_example(client)
-        microsoft_sharepoint_example(client)
-        salesforce_example(client)
-        sendme_example(client)
+        if HAVE_AMAZON_S3:
+            amazon_s3_example(client)
+        if HAVE_BOX:
+            box_example(client)
+        if HAVE_DROPBOX:
+            dropbox_example(client)
+        if HAVE_MICROSOFT_SHAREPOINT:
+            microsoft_sharepoint_example(client)
+        if HAVE_SALESFORCE:
+            salesforce_example(client)
+        if HAVE_SENDME:
+            sendme_example(client)
```

Commit:
fix auth _login, _refresh to use PAT, use access_tokens for all other requests, with proper auto rfresh

Commit:
use different auth and data endpoints for lexa, update base_client and async_base_client

Commit:
testing upgrades

Melabela committed:
docs: README.md - introduce Hippo in intro and add short section

- move 'Official Python SDK for' to before TOC, so these summary blurbs are part of file heading
   - keep existing 'Lexa' content, add 'Hippo' bullet with its own 'Perfect For' quote-block

- change heading 'Quick Start' to 'Lexa Quick Start'
- add heading 'Hippo Getting Started' with short section, and link to new placeholder doc

Melabela committed:
docs: rename some example/* files, to better show their alignment

Also update 'Examples' section in README.md:
- split links into two subsections & tables, by 'Lexa' & 'Document'
- update paths in 'Run Examples' code blurb

Melabela committed:
docs: split api-reference.md into 3, and move into docs/apis sub-folder

Make sub-folder docs/apis, as will add more files for Hippo.
Separate these from other docs, which feel more like guides.

Also split existing api-reference.md, into three files:
  - document.md, errors.md, lexa.md
makes it easier to find class info by api doc filename.

Melabela committed:
docs: update lexa.md api reference, with all list_* & parse_* functions

Melabela committed:
docs: add account.md api reference

Commit:
complete testing for base_client and async_base_client, will start ingestion refactor

Commit:
major structure refactor + tests

Melabela committed:
docs: add hippo.md api reference

Melabela committed:
docs: fill-in hippo getting started guide

Commit:
missed adding core files for client and constants

Commit:
fix: determine when to use data url and base url, all data is uploaded to data url

Commit:
feat: accept top k param for chat ask

Commit:
fix: send file size for large files and process with better limits

Commit:
feat: add support for include_retrieval and answer_options when submit_ask

Commit:
fix: increase default timeouts for query

Commit:
upgrade docs and testing for new v0.2.0 changes

Commit:
Merge docs_updates_ONTOP_api-ext-upg with preference for doc updates
